### PR TITLE
Add option to measure time spent inside calls to the CLR.

### PR DESCRIPTION
**Pull request by pkukol**
I suspect almost every line has illegal whitespace - and I apologize for that - plus there are a few small loose ends (like adding comments for folks changing the JIT/CLR API in the future), but I'd like folks to have plenty of time to take a look at this, so here goes. I'll wait for early comments while I clean it up and then post an update. Thanks for your patience.

PTAL @pgavlin @BruceForstall 
/cc @dotnet/jit-contrib

### CODE REVIEW

File: src/jit/compiler.cpp [hunk 1]
```diff
@@ -49,6 +49,13 @@ AssemblyNamesList2* Compiler::s_pAltJitExcludeAssembliesList            = nullpt
 #endif // ALT_JIT
 
 /*****************************************************************************/
+#ifdef _TARGET_AMD64_
+#include <intrin.h>
+inline bool _our_GetThreadCycles(unsigned __int64 *cycleOut){ *cycleOut = __rdtsc(); return true; }
```
Comment (janvorli) [inline src/jit/compiler.cpp]: To make this work xplat too, it seems you might want to use what we have in gc.cpp - see the get_cycle_count().

Comment (pkukol) [inline src/jit/compiler.cpp]: Oh, thanks @janvorli - you saved me time! x-plat is not the highest pri right off the bat (and, as all of my PR's, everything is off by default), but I'd like to make it work as widely as possible. I'll definitely try to steal / reuse the stuff you mention.

**Comment by pkukol** [comment]
@pgavlin @BruceForstall 
/cc @dotnet/jit-contrib
One more (pretty basic) thing - as soon as I have at least a few of the platforms either working with this stuff or explicitly mentioned / disabled in jit.h (so people don't try to use the unsupported configs) I'll temporarily push jit.h with this stuff enabled everywhere it can be, just to make sure all the builds still pass and so on. Once it works well enough that way I'll try to remember to turn it back off before merging the PR ;). I do realize this partly violates my tradition of never pushing code that's enabled by default - you never know where this might lead ...

**Comment by BruceForstall** [comment]
You could always do the "enable in jit.h" as a separate, later PR, marked "[DO NOT MERGE]", for example.

---
In reply to: [249635713](https://github.com/dotnet/coreclr/pull/7357#issuecomment-249635713) [](ancestors = 249635713)

File: src/jit/jitconfigvalues.h [hunk 2]
```diff
@@ -261,6 +261,8 @@ CONFIG_STRING(JitInlineReplayFile, W("JitInlineReplayFile"))
 CONFIG_INTEGER(JitInlinePolicyLegacy, W("JitInlinePolicyLegacy"), 0)
 CONFIG_INTEGER(JitInlinePolicyModel, W("JitInlinePolicyModel"), 0)
 
+CONFIG_INTEGER(JitCLRcallTimingInfo, W("JitCLRcallTimingInfo"), 0)
+
```
Comment (BruceForstall) [inline src/jit/jitconfigvalues.h]: The name of this strikes me as a bit weird, since in my mind, "CLR" means EE/VM + JIT + mscorlib + other base level stuff. How about just "JitCallTimingInfo"? Seems unique enough. Or, to be more explicit, JitItfCallTimingInfo or JitEECallTimingInfo.

File: src/jit/compiler.h [hunk 3]
```diff
@@ -960,12 +972,66 @@ struct CompTimeInfo
...
+#if MEASURE_CLRAPI_CALLS
+    // The following measures the time spent inside each individual CLR API call.
+    unsigned         m_allClrAPIcalls;
+    unsigned         m_perClrAPIcalls [API_ICorJitInfo_Names::API_COUNT];
+
+    unsigned __int64 m_allClrAPIcycles;
+    unsigned __int64 m_perClrAPIcycles[API_ICorJitInfo_Names::API_COUNT];
+
+    unsigned __int32 m_maxClrAPIcycles[API_ICorJitInfo_Names::API_COUNT];
+#endif // MEASURE_CLRAPI_CALLS
```
Comment (BruceForstall) [inline src/jit/compiler.h]: shouldn't this also be __int64?

File: src/jit/compiler.h [hunk 4]
```diff
@@ -1015,11 +1082,20 @@ class CompTimeSummaryInfo
...
+    unsigned __int64 m_CLRcallInvokes;  // CLR API invokes under current outer so far
+    unsigned __int64 m_CLRcallCycles;   // CLR API  cycles under current outer so far.
+    int              m_CLRcallAPInum;   // The enum/index of the current CLR API call (or -1).
+#endif
 #ifdef DEBUG
     Phases m_lastPhase; // The last phase that was completed (or (Phases)-1 to start).
 #endif
+#if MEASURE_CLRAPI_CALLS
+    static double    s_cyclesPerSec;    // Cached for speedier measurements
+#endif
```
Comment (BruceForstall) [inline src/jit/compiler.h]: couldn't you hoist this to the same `#if`, above?

File: src/jit/jitconfigvalues.h (hunk 2 above)
Comment (pkukol) [inline src/jit/jitconfigvalues.h]: That's pretty funny - it was JitEECallTimingInfo until this morning, when I changed a ton of names (thinking it was for the better). I shall revert this back to JitEECallTimingInfo, in the next push.

File: src/jit/compiler.h (hunk 3 above)
Comment (pkukol) [inline src/jit/compiler.h]: Actually, no - it's cycles inside the call which better fit in a 32-bit int. I'm trying to save memory because these arrays are not that small.

File: src/jit/compiler.h (hunk 4 above)
Comment (pkukol) [inline src/jit/compiler.h]: There are more things in this one in my local copy (which has many more changes), but when it's the sole survivor it's probably better to combine - will do.

File: src/jit/compiler.h [hunk 5]
```diff
@@ -960,12 +972,66 @@ struct CompTimeInfo
...
+    {
+        WrapICorJitInfo* wrap = nullptr;
+
+        if (JitConfig.JitCLRcallTimingInfo() != 0)
+        {
+            void* inst = alloc->allocateMemory(roundUp(sizeof(WrapICorJitInfo)));
+            if (inst != nullptr)
+            {
+                wrap = new(inst, jitstd::placement_t()) WrapICorJitInfo();
+
```
Comment (pkukol) [inline src/jit/compiler.h]: Makes sense - if I can remember now why I had to jump through such hoops ;).

File: src/jit/compiler.h [hunk 6]
```diff
@@ -1045,9 +1121,15 @@ class JitTimer
     // Ends the current phase (argument is for a redundant check).
     void EndPhase(Phases phase);
 
+#if MEASURE_CLRAPI_CALLS
+    // Start and end a timed CLR API call.
+    void CLRapiCallEnter(unsigned apix);
+    void CLRapiCallLeave(unsigned apix);
+#endif // MEASURE_CLRAPI_CALLS
```
Comment (BruceForstall) [inline src/jit/compiler.h]: In general, if we're using CamelCase (which we are...) then it seems the first letter of a word/name needs to be caps even if it follows an all-caps thing. Make, make the all-caps thing mixed, like is done in the C# standard. E.g., ClrApiCallEnter()

File: src/jit/compiler.hpp [hunk 7]
```diff
@@ -4425,6 +4425,30 @@ inline void Compiler::EndPhase(Phases phase)
 }
 
 /*****************************************************************************/
+#if MEASURE_CLRAPI_CALLS
+
+inline void Compiler::CLRapiCallEnter(unsigned apix)
+{
+    if (pCompJitTimer != NULL) pCompJitTimer->CLRapiCallEnter(apix);
+}
```
Comment (BruceForstall) [inline src/jit/compiler.hpp]: CC: NULL=>nullptr

File: src/jit/compiler.cpp [hunk 8]
```diff
@@ -49,6 +49,13 @@ AssemblyNamesList2* Compiler::s_pAltJitExcludeAssembliesList            = nullpt
 #endif // ALT_JIT
 
 /*****************************************************************************/
+#if defined(_TARGET_AMD64_) && defined(_MSC_VER)
+#include <intrin.h>
```
Comment (BruceForstall) [inline src/jit/compiler.cpp]: __rdtsc() also works for x86. So use `_TARGET_XARCH_`?

Comment (BruceForstall) [inline src/jit/compiler.cpp]: Actually, this isn't target-specific, it's host-specific. So probably: (defined(_HOST_X86_) || defined(_HOST_AMD64_))

---
In reply to: [80571468](https://github.com/dotnet/coreclr/pull/7357#discussion_r80571468) [](ancestors = 80571468)

**Comment by BruceForstall** [comment]
The biggest downside to this is that there are now three more places to update when adding a new JIT-EE interface method, in addition to the 5 or so places in SuperPMI that already need to get updated. I'm ok with this; the JIT-EE interface doesn't change that much (normally). In this case, however, since we expect all this code to be disabled normally, it's likely to atrophy.

**Comment by BruceForstall** [comment]
You need copyright headers in the new ICorJitInfo_API_*.h files.

**Comment by BruceForstall** [comment]
E.g., check out src\toolbox\superpmi\superpmi-shared\icorjitinfoimpl.h.

It has this note, which might be useful for yours as well:

```
// NOTE: this file is in exactly the same order, with exactly the same whitespace, as the ICorJitInfo
// interface declaration (with the "virtual" and "= 0" syntax removed). This is to make it easy to compare
// against the interface declaration.
```

or maybe you can use exactly the same file? (Not #include cross-directory, but copy?)

**Comment by pkukol** [comment]
@BruceForstall thanks for the reminder about header file comments; will do soon. I just deleted one of the header files so we're down to one, and yes it will atrophy no doubt. Unless we create an API definition that can be massaged into different formats (header files, etc), alas, I don't see any better way of doing it.

File: src/jit/compiler.cpp (hunk 8 above)
Comment (pkukol) [inline src/jit/compiler.cpp]: I copied (and tweaked) the stuff from gc.cpp suggested earlier - would you mind taking another look, @BruceForstall ? Thanks. And yes, you're absolutely right - we're measuring this code running on the host, it doesn't matter what it's targeting. Mea culpa.

**Comment by pkukol** [comment]
Hello, @BruceForstall @pgavlin @CarolEidt and everyone else ...
/cc @dotnet/jit-contrib
... I think that I've addressed all the issues / changes suggested by people and machines, so I'd appreciate it if you could give it a look and let me know if there's anything else left to do. Thanks!

File: src/jit/compphases.h [hunk 9]
```diff
@@ -91,6 +91,10 @@ CompPhaseNameMacro(PHASE_LINEAR_SCAN_RESOLVE,    "LSRA resolve",
 CompPhaseNameMacro(PHASE_GENERATE_CODE,          "Generate code",                  "CODEGEN",  false, -1)
 CompPhaseNameMacro(PHASE_EMIT_CODE,              "Emit code",                      "EMIT",     false, -1)
 CompPhaseNameMacro(PHASE_EMIT_GCEH,              "Emit GC+EH tables",              "EMT-GCEH", false, -1)
+
+#if MEASURE_CLRAPI_CALLS
+CompPhaseNameMacro(PHASE_CLR_API,                "CLR API calls",                  "CLR-API",  false, -1)
```
Comment (CarolEidt) [inline src/jit/compphases.h]: This might merit a simple comment indicating that this isn't really a "phase" in the usual sense.

File: tests/issues.targets [hunk 10]
```diff
@@ -40,6 +40,9 @@
         <ExcludeList Include="$(XunitTestBinBase)\GC\Scenarios\muldimjagary\muldimjagary\*">
             <Issue>3392</Issue>
         </ExcludeList>
+        <ExcludeList Include="$(XunitTestBinBase)\GC\Scenarios\DoublinkList\dlstack\*">
```
Comment (CarolEidt) [inline tests/issues.targets]: Are these changes really related to this PR?

**Comment by CarolEidt** [comment]
The JIT changes seem reasonable to me, though there are some methods without method headers. I didn't scrutinize the GC or VM changes.

---
In reply to: [250011070](https://github.com/dotnet/coreclr/pull/7357#issuecomment-250011070) [](ancestors = 250011070)

File: src/jit/compphases.h (hunk 9 above)
Comment (pkukol) [inline src/jit/compphases.h]: Good point @CarolEidt - I'll add a comment to that effect.

File: tests/issues.targets (hunk 10 above)
Comment (pkukol) [inline tests/issues.targets]: Must be from rebasing my branch; hopefully this won't cause any harm?

Comment (pkukol) [inline tests/issues.targets]: It was an artifact of a merge/rebase; hopefully I managed to revert the merge correctly (it's kind of a pain with git). Those files have disappeared from this page so hopefully all is well now ;).

File: src/jit/compphases.h (hunk 9 above)
Comment (pkukol) [inline src/jit/compphases.h]: Done.

**Comment by pkukol** [comment]
@CarolEidt - I added comments to the new methods, short descriptions for the trivial one-liner ones and nice multiline headers for the two non-trivial methods. Hope it's OK now? I simply tried to match what I found in the same file that didn't look like it was 20 years old ;).

**Comment by pkukol** [comment]
Does this LGTanyone in @dotnet/jit-contrib , please? If not, can you please let me know what I need to do? Thanks!

**Comment by CarolEidt** [comment]
There's actually a standard header, documented in [https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#94-function-header-comment](https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#94-function-header-comment). The rough guideline we've been using is that if you add a method you add a header, and if you make more than minor changes to an existing method you add a header if it's missing.

---
In reply to: [250039815](https://github.com/dotnet/coreclr/pull/7357#issuecomment-250039815) [](ancestors = 250039815)

**Comment by pkukol** [comment]
Thanks for explaining that, @CarolEidt ; I confess I was lazy and simply employed the "use common sense" and "look around and don't rock the local boat" approach and didn't realize that the result was unacceptable. Can you please take another look and if anything is still missing kindly tell me explicitly what needs to be done? Terms like "rough guideline" and "minor changes" (not to mention "trivial accessors and wrappers" used in the law book) are just a bit too vague for someone not in tune with the culture etc, and I'm pretty desperate for someone's approval at this point.

File: src/jit/compiler.cpp [hunk 11]
```diff
@@ -14,6 +14,7 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 #ifdef _MSC_VER
 #pragma hdrstop
 #endif // _MSC_VER
+#define __HORRID_HACK__
```
Comment (CarolEidt) [inline src/jit/compiler.cpp]: >HORRID [](start = 10, length = 6)

Was this meant to stay?

**Comment by CarolEidt** [comment]
LGTM; I'm assuming you'll squash all the commits.

---
In reply to: [250308554](https://github.com/dotnet/coreclr/pull/7357#issuecomment-250308554) [](ancestors = 250308554)

**Comment by CarolEidt** [comment]
:shipit:

File: src/jit/compiler.cpp (hunk 11 above)
Comment (pkukol) [inline src/jit/compiler.cpp]: Oh jeez that should never have made it into this branch, my apologies (too many balls in the air at the same time). I'll remove it immediately, of course - thanks for catching that!

Comment (CarolEidt) [inline src/jit/compiler.cpp]: Wish I could say I've never done anything like that ;-)
